src/qtt/algorithms/noise_analysis.py

```diff
@@ -0,0 +1,219 @@
+"""
+
+@author: eendebakpt
+"""
+
+# %%
+from typing import Tuple, Union, Optional, Dict, Any
+import numpy as np
+import matplotlib.pyplot as plt
+import scipy.signal
+import logging
+from qtt.algorithms.fitting import extract_lmfit_parameters
+from lmfit.models import LinearModel, Model
+from numpy.fft import irfft, rfftfreq
+from numpy.random import normal
+
+
+def power_law_model(frequency: Union[float, np.ndarray], A: float, alpha: float) -> np.ndarray:
+    """ Model for power law
+
+    This function implements the function y = A/x^alpha
+
+    Also see: https://en.wikipedia.org/wiki/Power_law
+    For alpha=1 we have the model for pink noise, see https://en.wikipedia.org/wiki/Pink_noise
+
+    Args:
+        frequency: Independent variable
+        A: Scalar factor of the model
+        alpha: Exponent of the model
+    Returns:
+        Calculated value of the model
+    """
+    return A / frequency**alpha
+
+
+def calculate_psd_welch(signal: np.ndarray, sample_rate: float, nperseg: Optional[int] = None) -> Tuple[np.ndarray, np.ndarray]:
+    """ Calculate the PSD using the Welch method
+
+    Args:
+        signal: Data for which to calculate the PSD
+        sample_rate: Sample rate of the data in Hz
+        nperseg: Passed to the scipy welch method
+    Returns:
+        Tuple of frequencies and PSD
+
+    """
+    frequencies, psd_welch = scipy.signal.welch(signal, sample_rate, nperseg=nperseg)
+    return frequencies, psd_welch
+
+
+def plot_psd(frequencies: np.ndarray, psd: np.ndarray, frequency_unit: str = 'Hz', signal_unit: str = 'V',
+             label: Optional[str] = None, fig: int = 1):
+    """ Plot calculated PSD """
+    if fig is not None:
+        Fig = plt.figure(fig)
+        plt.clf()
+        ax = Fig.add_subplot(111)
+    else:
+        ax = plt.gca()
+    ax.loglog(frequencies, psd, label=label, color='navy')
+    plt.xlabel(f'Frequency ({frequency_unit})')
+    plt.ylabel(f'Power spectral density ({signal_unit}$^2$/{frequency_unit})')
+
+
+def generate_pink_noise(sample_rate: float, number_of_samples: int, A: float = 1, alpha: float = 1):
+    """ Generate sampled data with 1/f noise
+
```

Comment:
Seems to be based on:
    Timmer, J. and Koenig, M.:
    On generating power law noise.
    Astron. Astrophys. 300, 707-710 (1995)

Reply:
Fair point. I added a reference